### PR TITLE
Touch Manager Gesture Cancellation

### LINKED ISSUE #673: Update TouchManager to support gesture cancellation
### Bug Report
As of v4.7, we are only updating the `SDLTouchType` enum with `CANCEL` and not updating the `SDLTouchManager` with the same. This needs to be done for developers to properly implement the changes.

### PR DESCRIPTION
**Pull request by NicoleYarroch**
Fixes #673 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Test cases were added to the `SDLTouchManagerSpec` class to test canceling tap, pan, and pinch gestures.

### Summary
Adds support for detecting a `SDLTouchType` of type `CANCEL`. The `SDLTouchManager` handles parsing the `SDLOnTouchType` notification from Core and issues a cancel delegate callback for pinch and pan gestures. If a single or double tap gesture is canceled, the tap is simply canceled. 

##### Enhancements
* Added support for detecting a `SDLTouchType` of type `CANCEL` in the `SDLTouchManager` class.
* Added two new optional methods to the `SDLTouchManagerDelegate`: 
    - `touchManager:manager pinchCanceledAtCenterPoint:point`
    - `touchManager:manager panningCanceledAtPoint:point`.

##### Bug Fixes
* Added `SDLRectangle.h` to both of the `.podspec` files.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: SmartDeviceLink/SDLTouchManager.m [hunk 1]
```diff
@@ -141,39 +153,42 @@ - (void)sdl_handleTouchBegan:(SDLTouch *)touch {
     _performingTouchType = SDLPerformingTouchTypeSingleTouch;
 
     switch (touch.identifier) {
-        case SDLTouchIdentifierFirstFinger:
+            case SDLTouchIdentifierFirstFinger:
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLTouchManager.m]: This indentation for the whole switch is incorrect, should look like:

```objc
switch (touch.identifier) {
    case SDLTouchIdentifierFirstFinger: {
        self.previousTouch = touch;
    } break;
    case SDLTouchIdentifierSecondFinger: {
        _performingTouchType = SDLPerformingTouchTypeMultiTouch;
        self.currentPinchGesture = [[SDLPinchGesture alloc] initWithFirstTouch:self.previousTouch secondTouch:touch];
        self.previousPinchDistance = self.currentPinchGesture.distance;
        if ([self.touchEventDelegate respondsToSelector:@selector(touchManager:pinchDidStartAtCenterPoint:)]) {
            [self.touchEventDelegate touchManager:self pinchDidStartAtCenterPoint:self.currentPinchGesture.center];
        }
    } break;
}
```

File: SmartDeviceLink/SDLTouchManager.m [hunk 2]
```diff
@@ -141,39 +153,42 @@ - (void)sdl_handleTouchBegan:(SDLTouch *)touch {
...
+ *  @param touch Gesture information
+ */
 - (void)sdl_handleTouchMoved:(SDLTouch *)touch {
     if ((touch.timeStamp - self.previousTouch.timeStamp) <= (self.movementTimeThreshold * NSEC_PER_USEC) || !self.isTouchEnabled) {
         return; // no-op
     }
 
     switch (self.performingTouchType) {
-        case SDLPerformingTouchTypeMultiTouch:
+            case SDLPerformingTouchTypeMultiTouch:
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLTouchManager.m]: Also indented incorrectly:

```objc
switch (self.performingTouchType) {
    case SDLPerformingTouchTypeMultiTouch: {
        switch (touch.identifier) {
            case SDLTouchIdentifierFirstFinger: {
                self.currentPinchGesture.firstTouch = touch;
            } break;
            case SDLTouchIdentifierSecondFinger: {
                self.currentPinchGesture.secondTouch = touch;
            } break;
        }

        if ([self.touchEventDelegate respondsToSelector:@selector(touchManager:didReceivePinchAtCenterPoint:withScale:)]) {
            CGFloat scale = self.currentPinchGesture.distance / self.previousPinchDistance;
            [self.touchEventDelegate touchManager:self
                     didReceivePinchAtCenterPoint:self.currentPinchGesture.center
                                        withScale:scale];
        }

        self.previousPinchDistance = self.currentPinchGesture.distance;
    } break;
    case SDLPerformingTouchTypeSingleTouch: {
        _performingTouchType = SDLPerformingTouchTypePanningTouch;
        if ([self.touchEventDelegate respondsToSelector:@selector(touchManager:panningDidStartAtPoint:)]) {
            [self.touchEventDelegate touchManager:self
                           panningDidStartAtPoint:touch.location];
        }
    } break;
    case SDLPerformingTouchTypePanningTouch: {
        if ([self.touchEventDelegate respondsToSelector:@selector(touchManager:didReceivePanningFromPoint:toPoint:)]) {
            [self.touchEventDelegate touchManager:self
                       didReceivePanningFromPoint:self.previousTouch.location
                                          toPoint:touch.location];
        }
    } break;

    case SDLPerformingTouchTypeNone: break;
}
```

File: SmartDeviceLink/SDLTouchManager.m [hunk 3]
```diff
@@ -183,43 +198,43 @@ - (void)sdl_handleTouchMoved:(SDLTouch *)touch {
...
-            switch (touch.identifier) {
-                case SDLTouchIdentifierFirstFinger:
-                    self.currentPinchGesture.firstTouch = touch;
-                    break;
-                case SDLTouchIdentifierSecondFinger:
-                    self.currentPinchGesture.secondTouch = touch;
-                    break;
-            }
-
+            case SDLPerformingTouchTypeMultiTouch:
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLTouchManager.m]: Also indented incorrectly:

```objc
switch (self.performingTouchType) {
    case SDLPerformingTouchTypeMultiTouch: {
        [self sdl_setMultiTouchFingerTouchForTouch:touch];
        if (self.currentPinchGesture.isValid) {
            if ([self.touchEventDelegate respondsToSelector:@selector(touchManager:pinchDidEndAtCenterPoint:)]) {
                [self.touchEventDelegate touchManager:self pinchDidEndAtCenterPoint:self.currentPinchGesture.center];
            }
            self.currentPinchGesture = nil;
        }
    } break;
    case SDLPerformingTouchTypePanningTouch: {
        if ([self.touchEventDelegate respondsToSelector:@selector(touchManager:panningDidEndAtPoint:)]) {
            [self.touchEventDelegate touchManager:self panningDidEndAtPoint:touch.location];
        }
    } break;
    case SDLPerformingTouchTypeSingleTouch: {
        if (self.singleTapTimer == nil) {
            // Initial Tap
            self.singleTapTouch = touch;
            [self sdl_initializeSingleTapTimerAtPoint:self.singleTapTouch.location];
        } else {
            // Double Tap
            [self sdl_cancelSingleTapTimer];

            NSUInteger timeStampDelta = touch.timeStamp - self.singleTapTouch.timeStamp;
            CGFloat xDelta = fabs(touch.location.x - self.singleTapTouch.location.x);
            CGFloat yDelta = fabs(touch.location.y - self.singleTapTouch.location.y);

            if (timeStampDelta <= self.tapTimeThreshold * NSEC_PER_USEC && xDelta <= self.tapDistanceThreshold && yDelta <= self.tapDistanceThreshold) {
                CGPoint centerPoint = CGPointCenterOfPoints(touch.location,
                                                            self.singleTapTouch.location);
                if ([self.touchEventDelegate respondsToSelector:@selector(touchManager:didReceiveDoubleTapAtPoint:)]) {
                    [self.touchEventDelegate touchManager:self
                               didReceiveDoubleTapAtPoint:centerPoint];
                }
            }

            self.singleTapTouch = nil;
        }
    } break;

    case SDLPerformingTouchTypeNone: break;
}
```

File: SmartDeviceLink/SDLTouchManager.m [hunk 4]
```diff
@@ -257,16 +276,80 @@ - (void)sdl_handleTouchEnded:(SDLTouch *)touch {
...
+    }
+
+    if (self.singleTapTimer != nil) {
+        // Cancel any ongoing single tap timer
+        [self sdl_cancelSingleTapTimer];
+        self.singleTapTouch = nil;
+    }
+
+    switch (self.performingTouchType) {
+            case SDLPerformingTouchTypeMultiTouch:
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLTouchManager.m]: Also indented incorrectly:

```objc
switch (self.performingTouchType) {
    case SDLPerformingTouchTypeMultiTouch: {
        [self sdl_setMultiTouchFingerTouchForTouch:touch];
        if (self.currentPinchGesture.isValid) {
            if ([self.touchEventDelegate respondsToSelector:@selector(touchManager:pinchCanceledAtCenterPoint:)]) {
                [self.touchEventDelegate touchManager:self pinchCanceledAtCenterPoint:self.currentPinchGesture.center];
            }
            self.currentPinchGesture = nil;
        }
    } break;
    case SDLPerformingTouchTypePanningTouch: {
        if ([self.touchEventDelegate respondsToSelector:@selector(touchManager:panningCanceledAtPoint:)]) {
            [self.touchEventDelegate touchManager:self panningCanceledAtPoint:touch.location];
        }
    } break;
    case SDLPerformingTouchTypeSingleTouch: // fallthrough
    case SDLPerformingTouchTypeNone: break;
}
```

File: SmartDeviceLink/SDLTouchManager.m [hunk 5]
```diff
@@ -257,16 +276,80 @@ - (void)sdl_handleTouchEnded:(SDLTouch *)touch {
...
 
+/**
+ *  Handles a CANCEL touch event sent by CORE. The CANCEL touch event is sent when a gesture is interrupted during a video stream. This can happen when a system dialog box appears on the screen, such as when the user is alerted about an incoming phone call. 
+ *
+ *  Pinch and pan gesture subscribers are notified if the gesture is canceled. Tap gestures are simply canceled without notification.
+ *
+ *  @param touch    Gesture information
+ */
+- (void)sdl_handleTouchCanceled:(SDLTouch *)touch {
+    if (!self.isTouchEnabled) {
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLTouchManager.m]: This already exists in the calling method and should not be repeated here or in the other sub-handlers.

File: SmartDeviceLink/SDLTouchManager.m [hunk 6]
```diff
@@ -257,16 +276,80 @@ - (void)sdl_handleTouchEnded:(SDLTouch *)touch {
...
+
+            case SDLPerformingTouchTypePanningTouch:
+            if ([self.touchEventDelegate respondsToSelector:@selector(touchManager:panningCanceledAtPoint:)]) {
+                [self.touchEventDelegate touchManager:self
+                                 panningCanceledAtPoint:touch.location];
+            }
+            break;
+
+            case SDLPerformingTouchTypeNone:
+            case SDLPerformingTouchTypeSingleTouch:
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLTouchManager.m]: Reverse this and the above case, they should be in the same order as other sub-handlers (this is done in the code provided in the above comment)

File: SmartDeviceLink/SDLTouchManager.m [hunk 7]
```diff
@@ -257,16 +276,80 @@ - (void)sdl_handleTouchEnded:(SDLTouch *)touch {
...
+
+    self.previousTouch = nil;
+    _performingTouchType = SDLPerformingTouchTypeNone;
+}
+
+- (void)sdl_setMultiTouchFingerTouchForTouch:(SDLTouch *)touch {
+    switch (touch.identifier) {
+        case SDLTouchIdentifierFirstFinger:
+            self.currentPinchGesture.firstTouch = touch;
+            break;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLTouchManager.m]: Add braces as in above cases

File: SmartDeviceLink/SDLTouchManager.m (hunk 1 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLTouchManager.m]: Fixed

File: SmartDeviceLink/SDLTouchManager.m (hunk 2 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLTouchManager.m]: Fixed

File: SmartDeviceLink/SDLTouchManager.m (hunk 3 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLTouchManager.m]: Fixed

File: SmartDeviceLink/SDLTouchManager.m (hunk 5 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLTouchManager.m]: Fixed. Removed from all sub-handlers

File: SmartDeviceLink/SDLTouchManager.m (hunk 4 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLTouchManager.m]: Fixed

File: SmartDeviceLink/SDLTouchManager.m (hunk 6 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLTouchManager.m]: Fixed

File: SmartDeviceLink/SDLTouchManager.m (hunk 7 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLTouchManager.m]: Fixed

**Pull request merged**